context-propagation-java5/src/main/java/nl/talsmasoftware/context/delegation/WrapperWithContext.java

```diff
@@ -24,36 +24,84 @@
+    private volatile ContextSnapshot _snapshot = null;
     protected final ContextSnapshotConsumer consumer;
 
     @Deprecated
     protected WrapperWithContext(ContextSnapshot snapshot, T delegate) {
         this(snapshot, delegate, null);
     }
 
-    protected WrapperWithContext(ContextSnapshot snapshot, T delegate, ContextSnapshotConsumer consumer) {
-        super(delegate);
-        this.snapshot = snapshot;
-        this.consumer = consumer;
+    protected WrapperWithContext(final ContextSnapshot snapshot, final T delegate, final ContextSnapshotConsumer consumer) {
+        this(new ContextSnapshotSupplier() {
+            public ContextSnapshot get() {
+                return snapshot;
+            }
+        }, delegate, consumer);
         if (snapshot == null) {
             throw new NullPointerException(String.format("No context snapshot provided to %s.", this));
         }
     }
 
+    /**
+     * Wraps the delegate and provides a context.
+     * <p>
+     * <strong>Note:</strong> <em>Make sure the supplier function does <strong>not</strong> obtain the context snapshot
+     * from any threadlocal storage! The wrapper is designed to propagate contexts from one thread to another.
+     * Therefore, the snapshot must be {@link nl.talsmasoftware.context.ContextManagers#createContextSnapshot() captured}
+     * in the source thread and {@link ContextSnapshot#reactivate() reactivated} in the target thread.
+     * If unsure, please use the
+     * {@link #WrapperWithContext(ContextSnapshot, Object, ContextSnapshotConsumer) constructor with snapshot} instead.</em>
+     *
+     * @param supplier The supplier for the context snapshot.
+     *                 This can be a straightforward 'holder' object or an ongoing background call.
+     *                 Please do <strong>not</strong> make this supplier function access any {@code ThreadLocal} value,
+     *                 as the wrapper is designed to propagate the snapshot from thread to thread!
+     * @param delegate The delegate object to be wrapped.
+     * @param consumer The consumer of a resulting context snapshot (optional, only required if the caller is interested in it).
+     * @see #WrapperWithContext(ContextSnapshot, Object, ContextSnapshotConsumer)
+     */
+    protected WrapperWithContext(ContextSnapshotSupplier supplier, T delegate, ContextSnapshotConsumer consumer) {
+        super(delegate);
+        this.supplier = supplier;
+        this.consumer = consumer;
+        if (supplier == null) {
+            throw new NullPointerException(String.format("No context snapshot supplier provided to %s.", this));
+        }
+    }
+
+    /**
+     * Calls the supplier for the context snapshot, making sure it is called only once for this wrapper instance.
+     *
+     * @return The snapshot value.
+     */
+    protected ContextSnapshot snapshot() {
+        // Note: Double-checked locking works fine in decent JDKs
+        if (_snapshot == null) synchronized (this) {
```

Comment:
Is it possible to use an AtomicReference::compareAndSet() instead?

Reply:
AtomicReference's `compareAndSet` requires an evaluated value, which means that in a race condition my guarantee that the `Supplier` is called only once is violated.  
This concern would be addressed by the `AtomicReference.updateAndGet` method, which was only added in Java 8 unfortunately.

Resorting to double-checked locking is the least complicated way I know of to guarantee the _only once_ promise we make regarding the context snapshot supplier.